File: elixir/anagram/lib/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    candidates |> Enum.filter(&anagram?(base, &1))
  end

  defp anagram?(base, candidate) do
    base_string = String.upcase(base)
    string = String.upcase(candidate)

    string != base_string and
      Enum.sort(String.to_charlist(string)) == Enum.sort(String.to_charlist(base_string))
  end
end

File: clojure/anagram/src/anagram.clj
(ns anagram
  (:require [clojure.string :as str]))

(defn anagram? [word tested-word]
  (let [uppercase-word (str/upper-case word) 
        upper-tested (str/upper-case tested-word)]
  (and 
    (not (= uppercase-word upper-tested)) 
    (= (sort uppercase-word) (sort upper-tested)))))

(defn anagrams-for [word prospect-list]
  (filter #(anagram? word %) prospect-list)
)
